lib/accessly/policy/base_option_1.rb

```diff
@@ -0,0 +1,86 @@
+module Accessly
+  module Policy
+    class BaseOption1
+
+      def initialize(actors)
+        @actors = actors
+      end
+
+      def self.actions(actions)
+        _actions.merge!(actions)
+        actions.each do |action, action_id|
+          _define_action_method(action, action_id)
+        end
+      end
+
+      def self.actions_on_objects(actions_on_objects)
+        _actions_on_objects.merge!(actions_on_objects)
+        actions_on_objects.each do |action, action_id|
+          _define_action_method(action, action_id)
+        end
+      end
+
+      def self.object_type
+        raise NotImplementedError.new("object_type must be implemented")
+      end
+
+      def object_type
+        self.class.object_type
+      end
+
+      def accessly_query
+        @_accessly_query ||= Accessly::Query.new(@actors)
+      end
+
+      private
+
+      def self._actions
+        @@_actions ||= {}
+      end
+
+      def self._actions_on_objects
+        @@_actions_on_objects ||= {}
+      end
+
+      def _actions
+        self.class._actions
+      end
+
+      def _actions_on_objects
+        self.class._actions_on_objects
+      end
+
+      def self._define_action_method(action, action_id)
+        unless method_defined?("#{action}?")
```

Comment:
One other potential option with this might be to overide "method_missing". It might end up being potentially easier to reason about than metaprogramming them? Something like:
```Ruby
def method_missing(method_name, *args)
  action = method_name.to_s.sub(/\?$/,"")
  if args.empty?
    action_id = actions[action]
    return super unless action_id
    accessly_query.can?(actions[action], object_type)
  else
    action_id = actions_on_object[action]
    return super unless action_id
    object = args.first
    accessly_query.can?(action_id, object.class, object.id)
  end
end
```
That's maybe a roughly what I'm picturing... and then if you defined anything more complex than a straight check, it would not be a missing method. If you didn't, then the this would catch that and make a basic check for that action?

Reply:
on thing that would make that even simpler is if we came up with a convention of naming the checker methods to distinguish an action from an action on... something like `#edit?` vs `#edit_on?` then we could just know which type it was from the name?